### DISCUSSION TITLE
Chain two or more components

### DISCUSSION THREAD

**Discussion by mbacicc** [django-commons/django-unicorn#193] [Ideas]
I was thinking that it could be cool to chain an event of the first component to trigger a specific chained component depending on the user action. I recently moved from a UIKit project to bootstrap, I don't miss the former framework except for the pre-built modals they have, for instance a prompt modal requesting for a single field, a Yes - No modal, something like .Net messageboxes. With that you can ask user confirmation before deleting content, you can manage file upload on a modal (maybe using django forms?) Having this possibility, I could create a modal component with pre-built user interface (maybe with same replaceable strings/infos) and send to the first component info about what component and what parameters to send with it depending on the buttons the user presses. I don't know if this could be useful in other use case scenarios, but it goes in the DRY direction, that I'm starting to strongly appreciate.
I don't know how easy could it be to be done, because it means that there would be different way to instantiate a component, unless there would be a parameter in the template to prevent __init__ execution for the chained component until it's really needed?

**Comment by adamghill** [comment]
This is an interesting idea. I need to think about how it could be done and what new hooks/APIs might need to be created to enable it.

Is this the root request: create a component programmatically with dynamic information?

It might make sense to create an issue with this.

**Reply by mbacicc** [reply]
I Adam,
I actually was not so sure how would it be the best way to implement it, like you suggested or by calling something like two nested component, but one of them more generic with some context info to model its behavior (like I was thinking, but it can have limitations with respect to a dynamically generated component).

Moreover, I was not sure I was imagining all possible case uses of the system, that's why I started the thread here.

I will explain what was my trigger for this idea:

![chained_unicorns (2)](https://user-images.githubusercontent.com/46646960/116514987-ffe5a680-a8cb-11eb-8db5-a3fe515a016a.png)
I'm doing a page to show the details of an electronic board, the same one of the PR model.

- I'm managing the details with a "edit-save" approach taken from your examples, but I also want to create the delete (with confirmation) and the add of a new board.
- For the delete I was thinking to a confirmation dialog modal, which was easy to be done in [UIkit](https://getuikit.com/docs/modal#modal-dialogs), and I wanted to use the same approach here. The confirmation dialog can be created once, but must execute the action wanted by the calling component, not its own.
- For the add I still didn't understood if to use standard Django forms in a new page, or maybe (if possible) use a "reusable" Unicorn modal form [like they do here](https://github.com/trco/django-bootstrap-modal-forms)


I can try to help in some way.

**Reply by adamghill** [reply]
Thanks for the detailed information and diagram. Did you ever get to a good solution for this? I _think_ a generic modal would be possible by passing in kwargs to a unicorn component. Those kwargs can even be template variables which get passed into the view which might be useful?

If you have any sample code or anything you are actively trying to figure out, let me know!
